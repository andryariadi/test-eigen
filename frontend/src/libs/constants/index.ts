export const headlinesCategories = ["general", "business", "entertainment", "health", "science", "sports", "technology"];

export const everythingLanguages = ["ar", "de", "en", "es", "fr", "he", "it", "nl", "no", "pt", "ru", "sv", "ud", "zh"];

export const newsSelection = ["everything", "top-headlines"];
