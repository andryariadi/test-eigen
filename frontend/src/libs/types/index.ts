interface ArticleSource {
  id: string | null;
  name: string;
}

export interface ArticleProps {
  source: ArticleSource;
  author: string | null;
  title: string;
  description: string | null;
  url: string;
  urlToImage: string | null;
  publishedAt: string;
  content: string | null;
}
