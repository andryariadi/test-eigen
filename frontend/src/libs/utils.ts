import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatNewsDate(isoDate: string): string {
  const date = new Date(isoDate);

  const options: Intl.DateTimeFormatOptions = {
    month: "long",
    day: "numeric",
    year: "numeric",
  };

  return date.toLocaleDateString("en-US", options);
}

export const stripHtmlTags = (html: string | null) => {
  if (typeof window === "undefined") {
    // Untuk server-side
    return html ? html.replace(/<[^>]*>?/gm, "") : "";
  } else {
    // Untuk client-side
    const doc = new DOMParser().parseFromString(html ?? "", "text/html");
    return doc.body.textContent || "";
  }
};
