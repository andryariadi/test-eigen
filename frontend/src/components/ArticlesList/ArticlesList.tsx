import { ArticleProps } from "@/libs/types";
import ArticleCard from "../ArticleCard/ArticleCard";
import { cn } from "@/libs/utils";

const ArticlesList = ({ articles, totalResult, isArticlesLength }: { articles: ArticleProps[]; totalResult?: number; isArticlesLength?: boolean }) => {
  console.log({ articles }, "<---articlesList");

  return (
    <section className={cn(`${isArticlesLength ? "px-4 md:px-10 lg:px-20 xl:px-32 pb-20" : "px-0 pb-0"} flex flex-col justify-start gap-5`)}>
      {/* Articles Length */}
      {isArticlesLength ? (
        <span className="text-slate-600 text-base font-normal">
          Showing: {articles.length} of {totalResult} articles
        </span>
      ) : (
        <h5 className="text-xl font-semibold text-slate-900">Other Articles</h5>
      )}

      {/* Articles Card List */}
      <div className="w-full grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-[24px]">
        {articles.map((article) => (
          <ArticleCard key={article.author} article={article} />
        ))}
      </div>
    </section>
  );
};

export default ArticlesList;
