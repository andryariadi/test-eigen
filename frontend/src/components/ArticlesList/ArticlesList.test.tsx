import { render, screen } from "@testing-library/react";
import ArticlesList from "./ArticlesList";
import { ArticleProps } from "@/libs/types";

// Mock ArticleCard component
jest.mock("../ArticleCard/ArticleCard", () => ({
  __esModule: true,
  default: ({ article }: { article: ArticleProps }) => (
    <div data-testid="article-card">
      <h3>{article.title}</h3>
      <p>{article.source.name}</p>
    </div>
  ),
}));

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

describe.skip("ArticlesList Component", () => {
  const mockArticles: ArticleProps[] = [
    {
      source: {
        id: "1",
        name: "Tech News",
      },
      author: "John Doe",
      title: "React Testing in 2024",
      description: "Learn how to test React components effectively",
      url: "https://example.com/react-testing",
      urlToImage: "/test1.jpg",
      publishedAt: "2024-01-01T00:00:00Z",
      content: "Lorem ipsum dolor sit amet...",
    },
  ];

  it("renders correctly with articles length info", () => {
    render(<ArticlesList articles={mockArticles} totalResult={15} isArticlesLength={true} />);

    // Test showing articles info
    const showingText = screen.getByText(`Showing: ${mockArticles.length} of 15 articles`);

    expect(showingText).toBeInTheDocument();
    expect(showingText).toHaveClass("text-slate-600");

    // Test article cards rendering
    const cards = screen.getAllByTestId("article-card");
    expect(cards).toHaveLength(mockArticles.length);

    // Test container classes
    const section = screen.getByTestId("region");
    expect(section).toHaveClass("px-4");
    expect(section).toHaveClass("pb-20");
  });

  it('renders correctly with "Other Articles" title', () => {
    render(<ArticlesList articles={mockArticles} isArticlesLength={false} />);

    const title = screen.getByText("Other Articles");

    expect(title).toBeInTheDocument();
    expect(title).toHaveClass("text-slate-900");
    expect(title).toHaveClass("text-xl");

    expect(screen.queryByText(/Showing:/)).not.toBeInTheDocument();

    // Test container classes
    const section = screen.getByTestId("region");
    expect(section).toHaveClass("px-0");
    expect(section).toHaveClass("pb-0");
  });

  it("applies correct grid layout classes", () => {
    render(<ArticlesList articles={mockArticles} />);

    const grid = screen.getByTestId("articles-grid");

    expect(grid).toHaveClass("grid-cols-1");
    expect(grid).toHaveClass("md:grid-cols-2");
    expect(grid).toHaveClass("lg:grid-cols-3");
    expect(grid).toHaveClass("gap-[24px]");
  });

  it("handles null values in article data", () => {
    const articlesWithNulls: ArticleProps[] = [
      {
        source: {
          id: null,
          name: "Null News",
        },
        author: null,
        title: "Article with Null Values",
        description: null,
        url: "https://example.com/null-article",
        urlToImage: null,
        publishedAt: "2024-01-03T00:00:00Z",
        content: null,
      },
    ];

    render(<ArticlesList articles={articlesWithNulls} />);

    expect(screen.getByText("Article with Null Values")).toBeInTheDocument();
    expect(screen.getByText("Null News")).toBeInTheDocument();
  });

  it('renders "No articles found" when totalResult is 0', () => {
    render(<ArticlesList articles={[]} totalResult={0} />);

    expect(screen.getByText("No articles found")).toBeInTheDocument();
  });
});
