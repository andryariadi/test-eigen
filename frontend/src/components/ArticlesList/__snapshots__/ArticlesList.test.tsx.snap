// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ArticlesList Component matches snapshot with full article data 1`] = `
<DocumentFragment>
  <section
    class="px-4 md:px-10 lg:px-20 xl:px-32 pb-20 flex flex-col justify-start gap-5"
    data-testid="region"
  >
    <span
      class="text-slate-600 text-base font-normal"
    >
      Showing: 1 of 15 articles
    </span>
    <div
      class="w-full grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-[24px]"
      data-testid="articles-grid"
    >
      <div
        data-testid="article-card"
      >
        <h3>
          React Testing in 2024
        </h3>
        <p>
          Tech News
        </p>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
